modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CreateTableCommand.java

```diff
@@ -0,0 +1,187 @@
+ */
+
+package org.apache.ignite.internal.catalog.commands;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.ignite.internal.catalog.CatalogParamsValidationUtils.ensureNoTableOrIndexExistsWithGivenName;
+import static org.apache.ignite.internal.catalog.CatalogParamsValidationUtils.validateColumnParams;
+import static org.apache.ignite.internal.catalog.CatalogParamsValidationUtils.validateIdentifier;
+import static org.apache.ignite.internal.catalog.commands.CatalogUtils.schemaOrThrow;
+import static org.apache.ignite.internal.catalog.commands.CatalogUtils.zoneOrThrow;
+import static org.apache.ignite.internal.util.CollectionUtils.nullOrEmpty;
+import static org.apache.ignite.lang.IgniteStringFormatter.format;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.apache.ignite.internal.catalog.Catalog;
+import org.apache.ignite.internal.catalog.CatalogValidationException;
+import org.apache.ignite.internal.catalog.descriptors.CatalogHashIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogSchemaDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogTableDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogZoneDescriptor;
+import org.apache.ignite.internal.catalog.storage.NewIndexEntry;
+import org.apache.ignite.internal.catalog.storage.NewTableEntry;
+import org.apache.ignite.internal.catalog.storage.ObjectIdGenUpdateEntry;
+import org.apache.ignite.internal.catalog.storage.UpdateEntry;
+
+/**
+ * A command that adds a new table to the catalog.
+ */
+public class CreateTableCommand extends AbstractCatalogCommand {
+    /** Schema name where this new table will be created. */
+    private final String schemaName;
+
+    /** Table name. */
+    private final String tableName;
+
+    /** Primary key columns. */
+    private final List<String> primaryKeyColumns;
+
+    /** Colocation columns. */
+    private final List<String> colocationColumns;
+
+    /** Columns. */
+    private final List<ColumnParams> columns;
+
+    /** Distribution zone name. */
+    private final String zoneName;
+
+    /**
+     * Constructs the object.
+     *
+     * @param tableName Name of the table to create. Should not be null or blank.
+     * @param schemaName Name of the schema to create table in. Should not be null or blank.
+     * @param primaryKeyColumns Name of columns which represent primary key.
+     *      Should be subset of columns in param `columns`.
+     * @param colocationColumns Name of the columns participating in distribution calculation.
+     *      Should be subset of the primary key columns.
+     * @param columns List of the columns containing by the table. There is should be at least one column.
+     * @param zoneName Name of the zone to create table in. Should not be null or blank.
+     * @throws CatalogValidationException if any of restrictions above is violated.
+     */
+    public CreateTableCommand(
```

Comment:
We usually restrict the ability to create directly with the constructor those classes that the builder creates :roll_eyes: :thinking:

Reply:
there is validation in contractor, thus it should be safe to create object directly. But anyways, I've reworked this part a bit, now constructor is private, so there is only one option to create command now